sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java

```diff
@@ -451,6 +518,27 @@ FormattedItemValue format(Object value) {
      * <p>Internal-only. Value objects can be safely cast to the expected Java type.
      */
     abstract FormattedItemValue format(Object value);
+
+    /**
+     * Determine whether the given value is compatible for the DisplayData type.
+     */
+    abstract boolean isCompatible(@Nullable Object value);
+
+    /**
+     * Infer the {@link Type} for the given object.
+     */
+    static Type inferFrom(@Nullable Object value) {
+      Set<Type> types = Sets.newHashSet(Type.values());
+      types.remove(STRING); // String is default
+
+      for (Type type : types) {
```

Comment:
I worry about this since if there are multiple types that `isCompatible` returns true for, then we're relying on the order of `Type.values()`. It might make sense to just write out the code here, so it is easier to see the logic used to do the inference.